rust-symcrypt/src/hash.rs

```diff
@@ -0,0 +1,186 @@
+use core::ffi::c_void;
+use std::mem;
+use std::ptr;
+use symcrypt_sys;
+
+pub const SHA256_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
+pub const SHA384_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
+
+pub struct Sha256State {
+    state: symcrypt_sys::SYMCRYPT_SHA256_STATE,
+}
+
+impl Sha256State {
+    pub fn new() -> Self {
+        let mut instance = Sha256State {
+            state: symcrypt_sys::SYMCRYPT_SHA256_STATE::default(),
+        };
+        unsafe {
+            symcrypt_sys::SymCryptSha256Init(&mut instance.state);
+        }
+        instance
+    }
+
+    pub fn append(&mut self, data: &[u8]) {
+        unsafe {
+            symcrypt_sys::SymCryptSha256Append(
+                &mut self.state,                    // pState
+                data.as_ptr(),                      // pbData
+                data.len() as symcrypt_sys::SIZE_T, // cbData
+            );
+        }
+    }
+
+    pub fn result(&mut self, result: &mut [u8; SHA256_RESULT_SIZE]) {
+        unsafe {
+            symcrypt_sys::SymCryptSha256Result(&mut self.state, result.as_mut_ptr());
+        }
+    }
+}
+
+impl Drop for Sha256State {
+    fn drop(&mut self) {
+        unsafe {
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.state) as *mut c_void,
+                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
+            )
+        }
+    }
+}
+
+pub fn sha256(data: &[u8], result: &mut [u8; SHA256_RESULT_SIZE]) {
```

Comment:
I'd consider having this function return `[u8; SHA256_RESULT_SIZE]` instead of taking it as an out parameter. Out parameters are somewhat uncommon in Rust, especially because you can use `Result` to both return a value and sucess/failure information (speaking of which, can `SymCryptSha256` fail?).

I think that depends on how people will use this though. There could be a cost from extra memory copies and such, but the optimizer is usually pretty good at eliminating those, I think.

Reply:
Since this is the low level wrapper, we intentionally wanted to avoid any potential extra allocations. This was the recommendation from the symcrypt experts.

Reply:
`[u8; SHA256_RESULT_SIZE]` is a value type, so there's no heap allocation to worry about. But, it does get allocated on the stack and that can result in lots of extra memcpy traffic if the optimizer can't figure it out, which seems like the bigger concern to me here.

Reply:
None of these crypto functions can fail.

We want to avoid potential extra copies. In some crypto patterns there might be a collection of nested hashes to be hashed. That's why we wanted the caller to provide the slice to be updated.

Reply:
After thinking further, I agree with Eric on not using uncommon out parameters. Instead should be the function's return value.

Reply:
Okay I can change it over, @mlindgren thoughts?

Reply:
Fixed via @eholk's commit

Reply:
> We want to avoid potential extra copies. In some crypto patterns there might be a collection of nested hashes to be hashed. That's why we wanted the caller to provide the slice to be updated.

What does this pattern look like in pseudo code? It might be worth adding both options to provide both ease of use and control when needed.

Reply:
Personally, I'd be inclined to go with the idiomatic solution of returning an array for now. Before we do any sort of public release, we should benchmark the wrappers and see what kind of overhead they're adding. If we find that Rust's optimizer is not smart enough to remove memcpy and this causes a perf problem, then we can change the API.

Reply:
Noted, I will return an array and make an issue to track benchmarking the wrappers.